code/modules/recycling/disposal-unit.dm

```diff
@@ -527,6 +527,160 @@
 		M.loc = src
 	flush()
 
+//Trap Door
+/obj/machinery/disposal/trapdoor
+	name = "trapdoor"
+	desc = "Almost like a door, but on floor."
+	density = 0
+	icon = 'icons/obj/atmospherics/pipes/trapdoor.dmi'
+	icon_state = "closed"
+	mode = 0
+	layer = 2.01
+	var/id = 1
+	var/auto_close = 1200
+	var/auto_close_on_mob = 100
+	var/sound_open = 'sound/machines/blast_door.ogg'
+	var/sound_close = 'sound/machines/blast_door.ogg'
+	var/open = FALSE
+
+/obj/machinery/disposal/trapdoor/proc/open()
+	if(flushing)
+		return
+	if(open)
+		return
+	flick("opening", src)
+	playsound(src.loc, sound_open, 100, 1)
+	icon_state = "open"
+	spawn(5)
+	for(var/mob/living/M in loc)
+		if(!M.floating)
+			M.loc = src
+			trap_flush()
+			if(auto_close_on_mob)
+				spawn(auto_close_on_mob)
+					close()
+	for(var/obj/item/O in loc)
+		if(!O.throwing || !O.anchored)
+			O.loc = src
+			trap_flush()
+	open = TRUE
```

Comment:
Trap_flush() proc checks if open is set to TRUE so this should be put immediately after the flushing and open checks.